Add a setURL method to BaseDataSource

I have an application that runs in JBoss and works across H2, MySQL, Oracle and, now, PostgreSQL.  I leverage XA data sources because I have a single transaction that spans multiple data sources.

So far I've been able to parameterize my configuration in such a way that I can easily switch between H2, MySQL and Oracle.  The configuration in question is the JBoss XA data source configuration where I specify the URL to the database.  For my previous databases the following configuration works:

``` xml
<xa-datasource-property name="URL">${database.url}</xa-datasource-property>
```

where `${database.url}` is replaced by my database URL.

Unfortunately when I ran this for PostgreSQL, JBoss threw an exception:

> Caused by: java.lang.NoSuchMethodException: Method setURL not found

Searching around I found this old [JBoss forum](https://developer.jboss.org/thread/26776?tstart=0) post where they instead recommend configuring the driver like this:

``` xml
<xa-datasource-property name="ServerName">hostname</xa-datasource-property>
<xa-datasource-property name="PortNumber">5432</xa-datasource-property>
<xa-datasource-property name="DatabaseName">dbname</xa-datasource-property>
```

This works, however I'd rather use the more succinct URL-style configuration.  The [BaseDataSource](https://github.com/pgjdbc/pgjdbc/blob/master/org/postgresql/ds/common/BaseDataSource.java) class currently has a setUrl method.  As with [H2](http://www.h2database.com/javadoc/org/h2/jdbcx/JdbcDataSource.html#setURL_String) and [MySQL](http://www.javafind.net/gate.jsp?q=/library/16/mysql-connector-java-5.1.29/apidocs/com/mysql/jdbc/jdbc2/optional/MysqlDataSource.html#setURL%28java.lang.String%29), I would like to have a setURL method in BaseDataSource that calls setUrl.

Pull requests are always welcome!

Dave Cramer

On 3 February 2015 at 18:18, Lonny Jacobson <notifications@github.com>
wrote:

> I have an application that runs in JBoss and works across H2, MySQL,
> Oracle and, now, PostgreSQL. I leverage XA data sources because I have a
> single transaction that spans multiple data sources.
>
> So far I've been able to parameterize my configuration in such a way that
> I can easily switch between H2, MySQL and Oracle. The configuration in
> question is the JBoss XA data source configuration where I specify the URL
> to the database. For my previous databases the following configuration
> works:
>
> <xa-datasource-property name="URL">${database.url}</xa-datasource-property>
>
> where ${database.url} is replaced by my database URL.
>
> Unfortunately when I ran this for PostgreSQL, JBoss threw an exception:
>
> Caused by: java.lang.NoSuchMethodException: Method setURL not found
>
> Searching around I found this old JBoss forum
> <https://developer.jboss.org/thread/26776?tstart=0> post where they
> instead recommend configuring the driver like this:
>
> <xa-datasource-property name="ServerName">hostname</xa-datasource-property>
> <xa-datasource-property name="PortNumber">5432</xa-datasource-property>
> <xa-datasource-property name="DatabaseName">dbname</xa-datasource-property>
>
> This works, however I'd rather use the more succinct URL-style
> configuration. The BaseDataSource
> <https://github.com/pgjdbc/pgjdbc/blob/master/org/postgresql/ds/common/BaseDataSource.java>
> class currently has a setUrl method. As with H2
> <http://www.h2database.com/javadoc/org/h2/jdbcx/JdbcDataSource.html#setURL_String>
> and MySQL
> <http://www.javafind.net/gate.jsp?q=/library/16/mysql-connector-java-5.1.29/apidocs/com/mysql/jdbc/jdbc2/optional/MysqlDataSource.html#setURL(java.lang.String)>,
> I would like to have a setURL method in BaseDataSource that calls setUrl.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pgjdbc/pgjdbc/issues/250>.
>

--- a/pgjdbc/src/main/java/org/postgresql/ds/common/BaseDataSource.java
+++ b/pgjdbc/src/main/java/org/postgresql/ds/common/BaseDataSource.java
@@ -1089,6 +1089,16 @@ public abstract class BaseDataSource implements CommonDataSource, Referenceable 
     }
   }
 
+  /**
+   * Sets properties from a {@link DriverManager} URL.
+   * Added to follow convention used in other DBMS.
+   *
+   * @param url properties to set
+   */
+  public void setURL(String url) {
+    setUrl(url);
+  }
+
   public String getProperty(String name) throws SQLException {
     PGProperty pgProperty = PGProperty.forName(name);
     if (pgProperty != null) {

--- a/pgjdbc/src/main/java/org/postgresql/ds/common/BaseDataSource.java
+++ b/pgjdbc/src/main/java/org/postgresql/ds/common/BaseDataSource.java
@@ -1076,6 +1076,15 @@ public abstract class BaseDataSource implements CommonDataSource, Referenceable 
   }
 
   /**
+   * Generates a {@link DriverManager} URL from the other properties supplied.
+   *
+   * @return {@link DriverManager} URL from the other properties supplied
+   */
+  public String getURL() {
+    return getUrl();
+  }
+
+  /**
    * Sets properties from a {@link DriverManager} URL.
    *
    * @param url properties to set

--- a/pgjdbc/src/test/java/org/postgresql/test/jdbc2/optional/OptionalTestSuite.java
+++ b/pgjdbc/src/test/java/org/postgresql/test/jdbc2/optional/OptionalTestSuite.java
@@ -17,6 +17,7 @@ import org.junit.runners.Suite;
 @RunWith(Suite.class)
 @Suite.SuiteClasses({SimpleDataSourceTest.class,
         SimpleDataSourceWithUrlTest.class,
+        SimpleDataSourceWithSetURLTest.class,
         ConnectionPoolTest.class,
         PoolingDataSourceTest.class,
         CaseOptimiserDataSourceTest.class})

--- a/pgjdbc/src/test/java/org/postgresql/test/jdbc2/optional/SimpleDataSourceWithSetURLTest.java
+++ b/pgjdbc/src/test/java/org/postgresql/test/jdbc2/optional/SimpleDataSourceWithSetURLTest.java
@@ -1,0 +1,29 @@
+/*
+ * Copyright (c) 2004, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.test.jdbc2.optional;
+
+import org.postgresql.jdbc2.optional.SimpleDataSource;
+import org.postgresql.test.TestUtil;
+
+/**
+ * Performs the basic tests defined in the superclass. Just adds the configuration logic.
+ */
+public class SimpleDataSourceWithSetURLTest extends BaseDataSourceTest {
+  /**
+   * Creates and configures a new SimpleDataSource using setURL method.
+   */
+  @Override
+  protected void initializeDataSource() {
+    if (bds == null) {
+      bds = new SimpleDataSource();
+      bds.setURL(String.format("jdbc:postgresql://%s:%d/%s?prepareThreshold=%d&logLevel=%s", TestUtil.getServer(),
+              TestUtil.getPort(), TestUtil.getDatabase(), TestUtil.getPrepareThreshold(), TestUtil.getLogLevel()));
+      bds.setUser(TestUtil.getUser());
+      bds.setPassword(TestUtil.getPassword());
+      bds.setProtocolVersion(TestUtil.getProtocolVersion());
+    }
+  }
+}

--- a/pgjdbc/src/test/java/org/postgresql/test/jdbc2/optional/SimpleDataSourceWithSetURLTest.java
+++ b/pgjdbc/src/test/java/org/postgresql/test/jdbc2/optional/SimpleDataSourceWithSetURLTest.java
@@ -5,6 +5,13 @@
 
 package org.postgresql.test.jdbc2.optional;
 
+import static org.junit.Assert.assertEquals;
+import static org.postgresql.Driver.parseURL;
+
+import java.util.Properties;
+
+import org.junit.Test;
+import org.postgresql.PGProperty;
 import org.postgresql.jdbc2.optional.SimpleDataSource;
 import org.postgresql.test.TestUtil;
 
@@ -19,11 +26,35 @@ public class SimpleDataSourceWithSetURLTest extends BaseDataSourceTest {
   protected void initializeDataSource() {
     if (bds == null) {
       bds = new SimpleDataSource();
-      bds.setURL(String.format("jdbc:postgresql://%s:%d/%s?prepareThreshold=%d&logLevel=%s", TestUtil.getServer(),
-              TestUtil.getPort(), TestUtil.getDatabase(), TestUtil.getPrepareThreshold(), TestUtil.getLogLevel()));
+      bds.setURL(String.format("jdbc:postgresql://%s:%d/%s?prepareThreshold=%d&loggerLevel=%s", TestUtil.getServer(), TestUtil.getPort(), TestUtil.getDatabase(), TestUtil.getPrepareThreshold(),
+              TestUtil.getLogLevel()));
       bds.setUser(TestUtil.getUser());
       bds.setPassword(TestUtil.getPassword());
       bds.setProtocolVersion(TestUtil.getProtocolVersion());
     }
+  }
+
+  @Test
+  public void testGetURL() throws Exception {
+    con = getDataSourceConnection();
+
+    String url = bds.getURL();
+    Properties properties = parseURL(url, null);
+
+    assertEquals(TestUtil.getServer(), properties.getProperty(PGProperty.PG_HOST.getName()));
+    assertEquals(Integer.toString(TestUtil.getPort()), properties.getProperty(PGProperty.PG_PORT.getName()));
+    assertEquals(TestUtil.getDatabase(), properties.getProperty(PGProperty.PG_DBNAME.getName()));
+    assertEquals(Integer.toString(TestUtil.getPrepareThreshold()), properties.getProperty(PGProperty.PREPARE_THRESHOLD.getName()));
+    assertEquals(TestUtil.getLogLevel(), properties.getProperty(PGProperty.LOGGER_LEVEL.getName()));
+  }
+  @Test
+  public void testSetURL() throws Exception {
+    initializeDataSource();
+
+    assertEquals(TestUtil.getServer(), bds.getServerName());
+    assertEquals(TestUtil.getPort(), bds.getPortNumber());
+    assertEquals(TestUtil.getDatabase(), bds.getDatabaseName());
+    assertEquals(TestUtil.getPrepareThreshold(), bds.getPrepareThreshold());
+    assertEquals(TestUtil.getLogLevel(), bds.getLoggerLevel());
   }
 }

--- a/pgjdbc/src/test/java/org/postgresql/test/jdbc2/optional/SimpleDataSourceWithSetURLTest.java
+++ b/pgjdbc/src/test/java/org/postgresql/test/jdbc2/optional/SimpleDataSourceWithSetURLTest.java
@@ -8,12 +8,13 @@ package org.postgresql.test.jdbc2.optional;
 import static org.junit.Assert.assertEquals;
 import static org.postgresql.Driver.parseURL;
 
-import java.util.Properties;
-
-import org.junit.Test;
 import org.postgresql.PGProperty;
 import org.postgresql.jdbc2.optional.SimpleDataSource;
 import org.postgresql.test.TestUtil;
+
+import org.junit.Test;
+
+import java.util.Properties;
 
 /**
  * Performs the basic tests defined in the superclass. Just adds the configuration logic.
@@ -47,6 +48,7 @@ public class SimpleDataSourceWithSetURLTest extends BaseDataSourceTest {
     assertEquals(Integer.toString(TestUtil.getPrepareThreshold()), properties.getProperty(PGProperty.PREPARE_THRESHOLD.getName()));
     assertEquals(TestUtil.getLogLevel(), properties.getProperty(PGProperty.LOGGER_LEVEL.getName()));
   }
+
   @Test
   public void testSetURL() throws Exception {
     initializeDataSource();

--- a/pgjdbc/src/test/java/org/postgresql/test/jdbc2/optional/SimpleDataSourceWithSetURLTest.java
+++ b/pgjdbc/src/test/java/org/postgresql/test/jdbc2/optional/SimpleDataSourceWithSetURLTest.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2004, PostgreSQL Global Development Group
+ * Copyright (c) 2017, PostgreSQL Global Development Group
  * See the LICENSE file in the project root for more information.
  */
 

